### DISCUSSION TITLE
Sidekiq / Redis configuration for Heroku

### DISCUSSION THREAD

**Discussion by kramerdog** [sidekiq/sidekiq#4987] [Q&A]
You say here [https://github.com/mperham/sidekiq/wiki/Using-Redis#using-an-env-variable] that configuring Sidekiq on Heroku is dead simple.

You then describe:
- using an ENV Variable REDIS_PROVIDER
- using an Initializer in config/initializers/sidekiq.rb

However, it is not clear to me if these are 2 different alternatives or are supposed to work in tandem?

By default my sidekiq.rb points to localhost.
Setting the REDIS_PROVIDER env variable on Heroku to REDISTOGO_URL achieves just that - 
ENV[REDIS_PROVIDER] = REDISTOGO_URL  (literally)
as you would expect, NOT the value of the environment variable REDISTOGO_URL .

Regardless, my Heroku app is still looking for Redis at localhost, presumably where sidekiq.rb is pointing it to.

I then came across [(https://github.com/mperham/sidekiq/wiki/Heroku#redis)] which seems to be suggesting the config in sidekiq.rb is what needs to be updated. I also note the use of the double ENV trick ENV[ENV[ which makes more sense but is contrary to the previous suggestion,

Neither suggestion make clear what I am supposed to do about application.rb which defines:
`config.cache_store = :redis_store, 'redis://localhost:6379'`

I have seen it suggested in the Rails Guide [(https://guides.rubyonrails.org/caching_with_rails.html)] that this should be "overridden" in production.rb with
`config.cache_store = :redis_cache_store, { url: ENV['REDIS_URL'] }`

which is EVEN MORE confusing as it mentions :redis_cache_store and NOT :redis_store, REDIS_URL and not REDIS_PROVIDER

Any words of wisdom would be much appreciated.

**Comment by mperham** [comment]
Show us your environment values, minus any passwords, etc.

**Comment by kramerdog** [comment]
Thanks for the response Mike.

Are you referring to the Environment Variables defined within Heroku, or otherwise what specifically?

Arent these the only relevant ones?
- REDIS_PROVIDER: REDISTOGO_URL
- REDISTOGO_URL: redis://redistogo:1b5cbc63d3f1d0936b96f175fc0c8638@sole.redistogo.com:9696/

I was kind of hoping you would answer my specific questions on where the configuration should be implemented of the available options:

1. Heroku (env variables)
2. Sidekiq initialiser
3. application.rb / production.rb (config.cache_store)
4. all of the above

I cant find anything that clarifies the inter-dependencies between these settings defined in the different locations, and the references I have given all seem to conflict with each other.

Thanks

**Comment by mperham** [comment]
The `redis_store` is a Rails cache thing, independent of Sidekiq.

You've configured Sidekiq correctly, you only need to set up the ENV variables in Heroku as you have. Your error is coming from Rails. The easiest solution is something like this:

```ruby
config.cache_store = :redis_cache_store, { url: (ENV['REDISTOGO_URL'] || 'redis://localhost:6379/0') }
```

**Comment by kramerdog** [comment]
OK, thanks. I had overlooked Rails making separate use of Redis to Sidekiq and requiring its own config.

So it looks like I do need both of:

1. Heroku (env variables)
2. Sidekiq initialiser

from the Sidekiq perspective.

You could consider making that a bit clearer in the documentation in the Sidekiq Github Wiki.

Thanks for your help.

**Comment by mperham** [comment]
Why do you need a Sidekiq initializer? I documented what you need: Env
variables. Your rails config is outside the scope of Sidekiq.

**Comment by kramerdog** [comment]
Maybe you are right - maybe I only thought I needed a Sidekiq initialiser because I was trying to fix the wrong thing.

There was a Sidekiq initialiser by default which I then changed. I'll try again without one at all. Thanks.

**Comment by kramerdog** [comment]
Sidekiq is definitely running now. Thanks! And I have installed the routes for the Sidekiq monitor.

So I can see now that I have 7 enqueued jobs but cant get any of them to actually run.

I tried to delete the queue using the Sidekiq monitor UI (which is great!) but got an error:

`Redis::CommandError (ERR unknown command 'unlink'):`

I believe that I have configured a worker dyno in Heroku as 

`bundle exec sidekiq -e production -C config/sidekiq.yml` which is confirmed in Heroku.

sidekiq.yml is the default as:

```
:verbose: false
:concurrency: 10
:timeout: 25

# Sidekiq will run this file through ERB when reading it so you can
# even put in dynamic logic, like a host-specific queue.
# http://www.mikeperham.com/2013/11/13/advanced-sidekiq-host-specific-queues/
:queues:
  - default
```

I saw in this post https://stackoverflow.com/questions/30233977/how-can-i-get-activejob-to-enqueue-jobs-in-sidekiq-on-heroku-when-called-from-th

to do:

` heroku ps:scale worker=1`

which I entered on the command line as I cant see anything about activating Dynos in the Heroku UI other than what I already did to Confirm the worker.

I am clearly still missing something?

I am very appreciative of all your help so far - I've got to be close to the end now!

**Reply by kramerdog** [reply]
Update:

1) I can confirm that removing all code from sidekiq.rb initialiser works as you suggested. Maybe that suggestion should be added to your notes because the default code which pointed to localhost did not work in Production?

2) It appears that I was trying to delete the default queue previously which gave the error message shown. Shouldnt this be handled more neatly though than crashing the application?

3) I can however delete individual jobs from the list presented when clicking the default queue.

4) Tried heroku redis FLUSHALL and emptying the queue

5) None of my enqueued jobs run though still.

**Reply by kramerdog** [reply]
Kept browsing and found another avenue...

heroku ps showed that the worker has crashed

Tried heroku restart and checked heroku logs....

`RuntimeError: You are connecting to Redis v3.2.12, Sidekiq requires Redis v4.0.0 or greater`

This is really confusing when:

https://devcenter.heroku.com/articles/heroku-redis-hobby-deprecation

suggests that newly added Redis add-ons are supposed to be Version 6.0

But I think that RedisToGo and Redis are 2 different things 

From https://stackoverflow.com/questions/57996477/rails-sidekiq-redis-on-heroku-something-wrong it seems that RedisToGo is stuck at version 3 of Redis and Sidekiq 6 wants version 4 at least so suggests:

gem 'sidekiq', '5.2.7'

Tried that, now nothing works.

Sidekiq monitor gives the error

`We're sorry, but something went wrong.`

Logs show:

`Redis::ConnectionError (Connection lost (ECONNRESET))`

and gives an error stack that refers to Redis 4.4.0 !

Nothing is ever easy in the world of Rails, very much the opposite of the marketing spiel!

**Reply by mperham** [reply]
I need you to back off on the passive aggressive tone. Remember I'm providing support to you for free.

All software development is painful at times. Every system can be frustrating to learn and deep understanding takes time and effort. You will fail over and over until you succeed.

I would dump Redis-to-Go and use Heroku Redis. Don't downgrade to Sidekiq 5 if possible. "redis-4.4.0" is the Redis client gem version, not the server version.

**Reply by kramerdog** [reply]
Mike

My sincere apologies if you were under the impression that I was making any commentary on yourself or the software. This was definitely not my intention and I was hoping that I was making it clear how grateful I have been for your assistance.

I was commenting on the Rails eco-system as a whole - while it has many good points, "quick and easy" are certainly not on the list with deployments bogged down by a seemingly relentless minefield of version mis-matches. My opinion only, based on ongoing frustrations and voiced in general and certainly not at you or anyone in particular.

The best point of the eco-system however, and to my mind its redeeming feature, is the extensive and generous community of individuals such as yourself who are willing to step in quickly and offer support.

So on to your suggestion to use Heroku Redis and not Redis-to-Go - we have basically concluded then that Sidekiq 6 is not compatible with Redis-to-Go and it is better to stick with Sidekiq 6 and switch to Heroku Redis? 

The good news is that I tried this out and it works! I just reset the REDIS_PROVIDER env variable in Heroku now to REDIS_URL and everything seems to be finally working. Thanks again.

**Reply by mperham** [reply]
Every framework is complex, "easy" is relative. 10 years ago it was far easier to do web development in Rails versus J2EE. Nevertheless, it still takes years to become comfortable and deeply knowledgable about Rails and its capabilities. Glad you got it working.

Redis-to-Go has not kept up with modern Redis. This means the service is increasingly irrelevant to everyone; no clue why they are letting it "die on the vine" like that.
